### ISSUE TITLE
InAppMessages Urban Airship 9.0.0

### ISSUE THREAD

**Issue by luispereira** [urbanairship/android-library#55]
In previous versions of Urban Airship I was fetching the pending messages in order to show them on my desired screen. My app have many screens, but I only want one screen to shown the in-apps notifications. However if the user is on one screen which is not supported to show in-apps it should delay them to when the user reaches a supported in-app screen. 

I did it in previous versions like this when the user reached the support in-app screen:
```
val inAppMessage = //get in app from AirshipReceiver (push received) and save them locally
if (inAppMessage != null) {
    //if there is a message waiting to be retrieved then put it back on urban, as urban ignore inapp  notification while application is re-initialized
    UAirship.shared().inAppMessageManager.pendingMessage = inAppMessage
}
//Shows the latest InAppMessage
UAirship.shared().inAppMessageManager.showPendingMessage(activity)
```

Is there any way to do this in current version 9.0.0, since the inAppMessageManager was changed and no longer have pendingMessages.

**Comment by rlepinski** [comment]
@luispereira By screen do you mean activity? You can exclude our in-app messaging from showing on activities by adding a [meta data entry](https://docs.urbanairship.com/guides/android-in-app-messaging/#excluding-activities) in the manifest. If its not as simple as that, you could use our [screen tracking](https://docs.urbanairship.com/platform/android/#android-screen-tracking) and extend the schedule to trigger off a certain screen:

```
        airship.getLegacyInAppMessageManager()
               .setScheduleBuilderExtender(new LegacyInAppMessageManager.ScheduleInfoBuilderExtender() {
                   @NonNull
                   @Override
                   public InAppMessageScheduleInfo.Builder extend(Context context, InAppMessageScheduleInfo.Builder builder, LegacyInAppMessage legacyMessage) {
                       Trigger trigger = Triggers.newScreenTriggerBuilder()
                                                 .setScreenName("some screen")
                                                 .setGoal(1)
                                                 .build();
                       builder.setTriggers(Collections.singletonList(trigger));

                       return builder;
                   }
               });
``` 

Which will trigger the in-app message once you call `UAirship.shared().getAnalytics().trackScreen("some screen");` if a message is pending.

**Comment by rlepinski** [comment]
You might want to take a look at https://docs.urbanairship.com/guides/android-in-app-messaging

**Comment by luispereira** [comment]
@rlepinski Thanks for the quick response, yes activities. I have in every activity that I don't want to show in-apps I use the manifest tag "com.urbanairship.push.iam.EXCLUDE_FROM_AUTO_SHOW". 
 
If the user is in one of these activities with that tag, the in-app will not be displayed (correctly). However, I want to keep those in-apps to show them when the user reaches a screen where the in-app can be shown. 

Previously, I was saving them locally when they reach `onPushReceived()` of `AirshipReceiver` and when the user reaches an activity that I want to show these in-apps I would give the in-app to urban so it could trigger the in-app manually

**Comment by rlepinski** [comment]
The meta data tag will not prevent the in-app message from displaying when it reaches the correct activity. Are you trying to workaround the 1 pending message limitation?

**Comment by luispereira** [comment]
Not quite, let me give you one example. 

Activity A (in-app screen - the only screen to show in-app to the user)
Activity B (Not an in-app screen) 

User goes form A to B and then receives an in-app, when the user goes back to activity A the user will not see the in-app anymore. From this moment on the in-app is lost.

**Comment by rlepinski** [comment]
Thats your desired behavior? Only allowing the in-app to be displayed the first time the user launches Activity A?

**Comment by luispereira** [comment]
Yes, once it is shown it can be dismissed for ever. Additionally if the user is on Activity B and in-app arrives he should see it on Activity A.

**Comment by rlepinski** [comment]
So only trigger the in-app on activity A? Thats the default behavior. As long as you have the exclude meta data flag I would expect that to work.

**Comment by luispereira** [comment]
However, when the user is on Activity B and receives the in-app, and then goes to activity A, Urban Airship is not showing them. Am I missing something?

**Comment by rlepinski** [comment]
Did you try it with SDK 9.0? I believe that should work just fine.

**Comment by luispereira** [comment]
Yes, I did try now with your example but now I'm receiving in every activity. Maybe is because I do a startActivityForResult() from activity A to the activity B ?

**Comment by rlepinski** [comment]
Could you provide verbose logs (make sure your app key/secret are not in them) that captures that behavior?

**Comment by luispereira** [comment]
This log stack is when I already am on Activity B and already done this `UAirship.shared().getAnalytics().trackScreen("some screen");` onCreate() of Activity A.

And this onPushReceived() from AirshipReceiver:

```
 airship.getLegacyInAppMessageManager()
               .setScheduleBuilderExtender(new LegacyInAppMessageManager.ScheduleInfoBuilderExtender() {
                   @NonNull
                   @Override
                   public InAppMessageScheduleInfo.Builder extend(Context context, InAppMessageScheduleInfo.Builder builder, LegacyInAppMessage legacyMessage) {
                       Trigger trigger = Triggers.newScreenTriggerBuilder()
                                                 .setScreenName("some screen")
                                                 .setGoal(1)
                                                 .build();
                       builder.setTriggers(Collections.singletonList(trigger));

                       return builder;
                   }
               });
```



`
02-15 16:09:26.322 9858-9858/com.mysample.app V/MyApp - UALib: GcmPushReceiver - Received intent: com.google.android.c2dm.intent.RECEIVE
02-15 16:09:26.325 9858-9858/com.mysample.app I/MyApp - UALib: Received push: {google.sent_time=1518207226900, com.urbanairship.push.PUSH_ID=bc0bdf90-0dd5-11e8-9b87-0242d5ff2226, com.urbanairship.push.CANONICAL_PUSH_ID=cc282763-fcf7-438f-a460-c8d82621eaed, com.urbanairship.push.ALERT= SOMETHING, com.urbanairship.in_app={"actions":{"on_click":{"^d":"myapp://example"}},"display":{"alert":"Something I want to show","type":"banner","position":"bottom"},"expiry":"2018-05-10T20:13:46"}, com.urbanairship.push.APID=c6461d56-036d-4ce1-8361-8a6825c5e0fc, com.urbanairship.actions={"^d":"myapp://example"}, from=GCM_SENDER_ID, google.message_id=0:1518207226905263%a4155678f9fd7ecd, google.ttl=2419200}
02-15 16:09:26.351 9858-9938/com.mysample.app V/MyApp - UALib: PreferenceDataStore - Saving preference: com.urbanairship.push.LAST_CANONICAL_IDS value: ["98a7bf1e-dc3f-48ea-80c4-d7b3d54e9bbb","74844241-545d-43af-a6eb-0d0b096bfbd2","1ac1c69c-e20f-40ba-b769-380445da1068","190fbe0c-23bd-4906-b694-02b8eab1eced","321a92ef-0a70-4a5d-a441-5a27b6351421","3b5b306b-6c22-444a-a2d3-69668c8c52d1","cc282763-fcf7-438f-a460-c8d82621eaed"]
02-15 16:09:26.357 9858-11485/com.mysample.app D/MyApp - UALib: Analytics disabled - ignoring event: push_arrived
02-15 16:09:26.376 9858-9858/com.mysample.app V/MyApp - UALib: ActionService - Received intent: com.urbanairship.actionservice.ACTION_RUN_ACTIONS startId: 1
02-15 16:09:26.379 9858-11485/com.mysample.app D/MyApp - UALib: LegacyInAppMessageManager - Received a Push with an in-app message.
02-15 16:09:26.382 9858-9938/com.mysample.app V/MyApp - UALib: PreferenceDataStore - Saving preference: com.urbanairship.push.iam.PENDING_MESSAGE_ID value: bc0bdf90-0dd5-11e8-9b87-0242d5ff2226
02-15 16:09:26.382 9858-9858/com.mysample.app I/MyApp - UALib: Action ^d will not be run. Registry predicate rejected the arguments: ActionArguments { situation: 1, value: "myapp:\/\/example", metadata: Bundle[{com.urbanairship.REGISTRY_ACTION_NAME=^d, com.urbanairship.PUSH_MESSAGE={google.sent_time=1518207226900, com.urbanairship.push.PUSH_ID=bc0bdf90-0dd5-11e8-9b87-0242d5ff2226, com.urbanairship.push.CANONICAL_PUSH_ID=cc282763-fcf7-438f-a460-c8d82621eaed, com.urbanairship.push.ALERT= SOMETHING, com.urbanairship.in_app={"actions":{"on_click":{"^d":"myapp://example"}},"display":{"alert":"Something I want to show","type":"banner","position":"bottom"},"expiry":"2018-05-10T20:13:46"}, com.urbanairship.push.APID=c6461d56-036d-4ce1-8361-8a6825c5e0fc, com.urbanairship.actions={"^d":"myapp://example"}, from=GCM_SENDER_ID, google.message_id=0:1518207226905263%a4155678f9fd7ecd, google.ttl=2419200}}] }
02-15 16:09:26.382 9858-9946/com.mysample.app V/MyApp - UALib: AutomationEngine - Cancelled schedule group: 65e12850-0dd5-11e8-a8bc-02424501ac67
02-15 16:09:26.386 9858-9858/com.mysample.app D/MyApp - UALib: LegacyInAppMessageManager - Pending in-app message replaced.
02-15 16:09:26.387 9858-9858/com.mysample.app D/MyApp - UALib: Analytics disabled - ignoring event: in_app_resolution
02-15 16:09:26.390 9858-11485/com.mysample.app V/MyApp - UALib: NotificationIdGenerator - Incrementing notification ID count
02-15 16:09:26.394 9858-11485/com.mysample.app V/MyApp - UALib: NotificationIdGenerator - Notification ID: 1007
02-15 16:09:26.411 9858-11485/com.mysample.app I/MyApp - UALib: Posting notification: Notification(pri=0 contentView=null vibrate=default sound=default defaults=0x3 flags=0x10 color=0x00000000 vis=PUBLIC semFlags=0x0 semPriority=0) id: 1007 tag: null
02-15 16:09:26.411 9858-9946/com.mysample.app V/MyApp - UALib: AutomationEngine - Scheduled entries: [com.urbanairship.iam.InAppMessageSchedule@2b0197]
02-15 16:09:26.423 9858-11487/com.mysample.app D/MyApp - UALib: InAppMessageManager - Preparing schedule: ee24cebb-eece-4dcb-99b7-8252d9c66809
02-15 16:09:26.424 9858-11487/com.mysample.app D/MyApp - UALib: InAppMessageManager - Scheduled message prepared for display: ee24cebb-eece-4dcb-99b7-8252d9c66809
02-15 16:09:26.434 9858-9858/com.mysample.app D/MyApp - UALib: CommanderAirshipReceiver - Received intent with action: com.urbanairship.push.RECEIVED
02-15 16:09:26.441 9858-9858/com.mysample.app D/MyApp - UALib: InAppMessageManager - Displaying schedule: ee24cebb-eece-4dcb-99b7-8252d9c66809
02-15 16:09:26.445 9858-9858/com.mysample.app I/MyApp - UALib: BannerAdapter - Displaying in-app message.
02-15 16:09:26.445 9858-9858/com.mysample.app V/MyApp - UALib: InAppMessagingManager - Message displayed with scheduleId: ee24cebb-eece-4dcb-99b7-8252d9c66809
02-15 16:09:26.446 9858-9858/com.mysample.app D/MyApp - UALib: Analytics disabled - ignoring event: in_app_display
02-15 16:09:26.465 9858-9858/com.mysample.app D/TextView: setTypeface with style : 0
02-15 16:09:26.466 9858-9858/com.mysample.app D/TextView: setTypeface with style : 0
02-15 16:09:26.468 9858-9858/com.mysample.app D/TextView: setTypeface with style : 0
02-15 16:09:26.471 9858-9858/com.mysample.app V/MyApp - UALib: InAppMessagingManager - Requesting display lock for schedule: ee24cebb-eece-4dcb-99b7-8252d9c66809
02-15 16:09:26.471 9858-9858/com.mysample.app V/MyApp - UALib: InAppMessagingManager - Schedule already obtained lock.
02-15 16:09:26.482 9858-11488/com.mysample.app V/MyApp - UALib: AirshipGcmTaskService - Running job: JobInfo{action=ACTION_RICH_PUSH_USER_UPDATE, id=7, extras='{"EXTRA_FORCEFULLY":false}', airshipComponentName='com.urbanairship.richpush.RichPushInbox', isNetworkAccessRequired=false, initialDelay=0, persistent=false}
02-15 16:09:26.482 9858-11488/com.mysample.app V/MyApp - UALib: AirshipGcmTaskService - Waiting for jobInfo: JobInfo{action=ACTION_RICH_PUSH_USER_UPDATE, id=7, extras='{"EXTRA_FORCEFULLY":false}', airshipComponentName='com.urbanairship.richpush.RichPushInbox', isNetworkAccessRequired=false, initialDelay=0, persistent=false} to complete.
02-15 16:09:26.483 9858-9951/com.mysample.app V/MyApp - UALib: Job - Finished: JobInfo{action=ACTION_RICH_PUSH_USER_UPDATE, id=7, extras='{"EXTRA_FORCEFULLY":false}', airshipComponentName='com.urbanairship.richpush.RichPushInbox', isNetworkAccessRequired=false, initialDelay=0, persistent=false} with result: 0
02-15 16:09:26.484 9858-11488/com.mysample.app V/MyApp - UALib: AirshipGcmTaskService - JobInfo finished: JobInfo{action=ACTION_RICH_PUSH_USER_UPDATE, id=7, extras='{"EXTRA_FORCEFULLY":false}', airshipComponentName='com.urbanairship.richpush.RichPushInbox', isNetworkAccessRequired=false, initialDelay=0, persistent=false}
02-15 16:09:41.488 9858-9858/com.mysample.app V/MyApp - UALib: InAppMessagingManager - Message finished. ScheduleID: ee24cebb-eece-4dcb-99b7-8252d9c66809
02-15 16:09:41.490 9858-9858/com.mysample.app D/MyApp - UALib: Analytics disabled - ignoring event: in_app_resolution
02-15 16:09:41.493 9858-11487/com.mysample.app D/MyApp - UALib: InAppMessageManager - Schedule finished: ee24cebb-eece-4dcb-99b7-8252d9c66809
02-15 16:09:41.497 9858-9946/com.mysample.app V/MyApp - UALib: AutomationEngine - Schedule finished: ee24cebb-eece-4dcb-99b7-8252d9c66809
02-15 16:09:41.497 9858-9946/com.mysample.app V/MyApp - UALib: AutomationEngine - Deleting schedule: ee24cebb-eece-4dcb-99b7-8252d9c66809
`

**Comment by luispereira** [comment]
I actually did this by doing some reflection (since I did not want to break any kind of copyright law).
On AirshipReceiver when in-app is received in onPushReceived() I save the push information and add a listener to keep tracking if the push was or not displayed, something like this:

```
 @Override
    protected void onPushReceived(@NonNull Context context, @NonNull PushMessage message, boolean notificationPosted) {
         //here I do something to check if is in-app 
         UAirship.shared().getInAppMessagingManager().addListener(new InAppMessageListener() {
            @Override
            public void onMessageDisplayed(@NonNull String s, @NonNull InAppMessage inAppMessage) {
                 //here remove the local saved inappmessage
            }

            @Override
            public void onMessageFinished(@NonNull String s, @NonNull InAppMessage inAppMessage, @NonNull ResolutionInfo resolutionInfo) {

            }
        });
        //then save in-app locally 
```

Then on my screen which supports in-app I did something like this:

```
    @Override
    protected void onResume() {
        super.onResume();
       //get in-app from local storage, then apply reflection to build a schedule in-app 
        InAppMessageScheduleInfo inAppMessageScheduleInfo =  uaReflectionCreator(context, legacyInAppMessage);

        if(inAppMessageScheduleInfo != null){
             UAirship.shared().getInAppMessagingManager().scheduleMessage(inApp);
        }
   }

```

I don't know if there is another good option for this, if there is please tell me

**Comment by rlepinski** [comment]
First of all, the listeners and extenders you are adding should be done after takeOff (either in autopilot or application onCreate). The receiver gets called multiple times after the messages are processed, so its possible those things will not be set by the time the message is processed.

None of the extra code should be necessary. Your use case is perfectly valid and supported. I bet its either a misunderstanding of how the feature works or some other config issue. Instead of trying to work around whatever issue you are having, could your provide us a clear way to reproduce? A sample app?

The logs you posted show a normal flow of an in-app message. I am not even sure what the main problem is. Are you trying to queue up more than 1 in-app message at a time?

The code you posted with reflection is dangerous and not needed. Are you calling public methods using reflection? You risk us breaking your application on any update.

**Comment by luispereira** [comment]
There are many use cases and features that I want to implement using the approach I said, however you can reproduce this with the following steps:
 - Go any activity (or simply do not start app at all) which does not allow in-app messages (with a tag on manifest com.urbanairship.push.iam.EXCLUDE_FROM_AUTO_SHOW) 
 - Send a in-app to your device and wait to be received
 - Go back to an activity which allows in-apps 

Expected result: In-app should be received on the activity witch allows in-apps 

Actual result: In-app is not received at all and will never be received under these circunstances

**Comment by rlepinski** [comment]
I think I tracked the issue down. Ill work on a release today. Thanks for reporting the issue and working with me to understand the problem. Once released, your use case should be fully supported without any extra app code, just the metadata in the manifest.

**Comment by rlepinski** [comment]
Fixed in 9.0.1
